stdlib/runtime/malloc.gr

```diff
@@ -0,0 +1,275 @@
+ *
+ *   Grain             C Equivalent
+ *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *   let ptr      ===  union header *ptr
+ *   getNext(ptr) ===  ptr->s.ptr
+ *   getSize(ptr) ===  ptr->s.size
+ */
+
+/**
+ * Pointer to the start of the free list. This is always a multiple of
+ * 8, with the exception of its initial value (used as a sentinel).
+ */
+let mut freePtr = 1n
+
+/**
+ * Size (in bytes) of entries in the free list.
+ */
+let mallocHeaderSize = 8n
+
+/**
+ * log_2(mallocHeaderSize) (multiplication by the header
+ * size is equivalent to left-shifting by this amount)
+ */
+let logMallocHeaderSize = 3n
+
+/**
+ * The current size (in bytes) of the heap.
+ */
+let mut heapSize = 0n
+
+let _RESERVED_RUNTIME_SPACE = 4096n;
+
+let getNext = (ptr: WasmI32) => {
+  WasmI32.load(ptr, 0n)
+}
+
+let setNext = (ptr: WasmI32, val: WasmI32) => {
+  WasmI32.store(ptr, val, 0n)
+}
+
+let getSize = (ptr: WasmI32) => {
+  WasmI32.load(ptr, 4n)
+}
+
+let setSize = (ptr: WasmI32, val: WasmI32) => {
+  WasmI32.store(ptr, val, 4n)
+}
+
+/**
+ * Requests that the heap be grown by the given number of bytes.
+ *
+ * @param nbytes: WasmI32 - The number of bytes requested
+ * @return WasmI32 - If unsuccessful, -1. Otherwise, the pointer to the beginning of the extended region.
+ */
+let growHeap = (nbytes: WasmI32) => {
+  let mut reqSize = 0n
+  let mut reqResult = 0n
+  let mut origSize = heapSize
+
+  // If the size has not been initialized, do so.
+  if (heapSize == 0n) {
+    heapSize = initialHeapSize - _RESERVED_RUNTIME_SPACE
+    if (nbytes > heapSize) {
+      // More bytes requested than the initial heap size,
+      // so we need to request more anyway.
+      reqSize = nbytes - heapSize
+      reqSize = reqSize >> 16n
+      reqSize = reqSize + 1n
+      reqResult = extGrowHeap(reqSize)
+      if (reqResult == -1n) {
+        -1n
+      } else {
+        heapSize += reqSize << 16n
+        _RESERVED_RUNTIME_SPACE
+      }
+    } else {
+      _RESERVED_RUNTIME_SPACE
+    }
+  } else {
+    // The size has already been initialized, so call the external function.
+    reqSize = nbytes
+    reqSize = reqSize >> 16n
+    reqSize = reqSize + 1n
+    reqResult = extGrowHeap(reqSize)
+    if (reqResult == -1n) {
+      -1n
+    } else {
+      heapSize += reqSize << 16n
+      reqResult
+    }
+  }
+}
+
+/**
+ * Frees the given allocated pointer.
+ *
+ * @param ap: WasmI32 - The pointer to free
+ */
+export let free = (ap: WasmI32) => {
+  // Let memory manager know we're freeing this pointer
+  extFree(ap)
+
+  let mut blockPtr = ap - 16n // 8 bytes for malloc header + 8 bytes for GC header
+  let mut p = freePtr
+
+  // Edge case: for the first free (called by morecore), the free pointer
+  // is actually already pointing to this node, so we don't do anything.
+  if (blockPtr != freePtr) {
+    // Find the location to insert this block into the free list
+    for ( ; !((blockPtr > p) && (blockPtr < getNext(p))); p = getNext(p) ) {
```

Comment:
I would prefer not to do that in this PR—this file is supposed to be as 1:1 as possible to the original C implementation to avoid mistakes. Happy to make changes in the future, but I don't want to change it too much right now.

Reply:
Okay. Please open an issue 🙇

Reply:
https://github.com/grain-lang/grain/issues/545